File: src/parser/parser.cpp
#include <iostream>
#include <vector>
#include <memory>

#include "./parser.hpp"

std::vector<std::shared_ptr<ASTNode>> Parser::parse()
{
    std::vector<std::shared_ptr<ASTNode>> statements{};

    while (!isAtEnd())
    {
        try
        {
            auto stmt = expression(0);
            if (stmt)
            {
                statements.push_back(std::move(stmt));
            }
        }
        catch (const std::exception &e)
        {
            std::cerr << "Erro o parser: " << e.what() << '\n';
            break;
        }
    }
    return statements;
}

std::shared_ptr<ASTNode> Parser::expression(int rbp)
{
    const Token &t = advance();
    auto left = nud(t);
    if (!left)
    {
        return nullptr;
    }

    while (!isAtEnd() && rbp < getPrecedence(peek().type))
    {
        const Token &opToken = advance();
        auto right = led(left, opToken);
        if (!right)
        {
            break;
        }
        left = right;
    }
    return left;
}

const Token &Parser ::peek() const
{
    return tokens[pos];
}

const Token &Parser::consume(TokenType t, std::string message)
{
    if (peek().type == t)
    {
        pos++;
        return tokens[pos - 2];
    }
    // ERR
    std::cerr << "Error: " + message + '\n';
    std::cerr << "Value: " + peek().lexeme + '\n';
    exit(-1);
}

bool Parser::isAtEnd()
{
    return pos >= tokens.size() || tokens[pos].type == TokenType::END_OF_FILE;
}

const Token &Parser::advance()
{
    return tokens[pos++];
}

std::shared_ptr<ASTNode> Parser::nud(const Token &token)
{
    if (token.type == TokenType::END_OF_FILE)
    {
        return nullptr;
    }

    switch (token.type)
    {
    case TokenType::NUMBER:
        return std::make_shared<NumberNode>(std::stod(token.lexeme), token.loc);
    case TokenType::STRING:
        return std::make_shared<StringNode>(token.lexeme, token.loc);
    case TokenType::IDENTIFIER:
        return std::make_shared<IdentifierNode>(token.lexeme, token.loc);
    case TokenType::NEW:
        return parse_new_decl();
    default:
        throw std::runtime_error("Token inválido no início da expressão: " + token.lexeme);
    }
}

std::shared_ptr<ASTNode> Parser::led(std::shared_ptr<ASTNode> left, const Token &token)
{
    int precedence = getPrecedence(token.type);
    switch (token.type)
    {
    case TokenType::PLUS:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("+", left, right, token.loc);
    }
    case TokenType::MINUS:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("-", left, right, token.loc);
    }
    case TokenType::STAR:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("*", left, right, token.loc);
    }
    case TokenType::SLASH:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("/", left, right, token.loc);
    }
    case TokenType::PERCENT:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("%", left, right, token.loc);
    }
    case TokenType::POWERING:
    {
        // Right-associative: lower the precedence for the recursive call.
        auto right = expression(precedence - 1);
        return std::make_shared<BinaryOpNode>("**", left, right, token.loc);
    }
    case TokenType::OR:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("or", left, right, token.loc);
    }
    case TokenType::AND:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("and", left, right, token.loc);
    }
    case TokenType::EQUAL_EQUAL:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("==", left, right, token.loc);
    }
    case TokenType::BANG_EQUAL:
    case TokenType::NOT_EQUAL:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("!=", left, right, token.loc);
    }
    case TokenType::LESS:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("<", left, right, token.loc);
    }
    case TokenType::LESS_EQUAL:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("<=", left, right, token.loc);
    }
    case TokenType::GREATER:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>(">", left, right, token.loc);
    }
    case TokenType::GREATER_EQUAL:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>(">=", left, right, token.loc);
    }
    case TokenType::AMPERSAND:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("&", left, right, token.loc);
    }
    case TokenType::PIPE:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("|", left, right, token.loc);
    }
    case TokenType::CARET:
    {
        auto right = expression(precedence);
        return std::make_shared<BinaryOpNode>("^", left, right, token.loc);
    }
    case TokenType::ASSIGN:
    {
        // Right-associative: lower the precedence for the recursive call.
        auto right = expression(precedence - 1);
        return std::make_shared<BinaryOpNode>("=", left, right, token.loc);
    }
    case TokenType::PLUS_ASSIGN:
    {
        auto right = expression(precedence - 1);
        return std::make_shared<BinaryOpNode>("+=", left, right, token.loc);
    }
    case TokenType::MINUS_ASSIGN:
    {
        auto right = expression(precedence - 1);
        return std::make_shared<BinaryOpNode>("-=", left, right, token.loc);
    }
    case TokenType::STAR_ASSIGN:
    {
        auto right = expression(precedence - 1);
        return std::make_shared<BinaryOpNode>("*=", left, right, token.loc);
    }
    case TokenType::SLASH_ASSIGN:
    {
        auto right = expression(precedence - 1);
        return std::make_shared<BinaryOpNode>("/=", left, right, token.loc);
    }
    case TokenType::PERCENT_ASSIGN:
    {
        auto right = expression(precedence - 1);
        return std::make_shared<BinaryOpNode>("%=", left, right, token.loc);
    }
    default:
        std::cerr << "Token inválido no início da expressão: " << token.lexeme << std::endl;
        std::exit(1);
    }
}

int Parser::getPrecedence(TokenType type)
{
    switch (type)
    {
    // Assignment operators (lowest precedence, right-associative)
    case TokenType::ASSIGN:
    case TokenType::PLUS_ASSIGN:
    case TokenType::MINUS_ASSIGN:
    case TokenType::STAR_ASSIGN:
    case TokenType::SLASH_ASSIGN:
    case TokenType::PERCENT_ASSIGN:
        return 1;

    // Logical operators
    case TokenType::OR:
        return 2;
    case TokenType::AND:
        return 3;

    // Equality operators
    case TokenType::EQUAL_EQUAL:
    case TokenType::BANG_EQUAL:
    case TokenType::NOT_EQUAL:
        return 7;

    // Relational operators
    case TokenType::LESS:
    case TokenType::LESS_EQUAL:
    case TokenType::GREATER:
    case TokenType::GREATER_EQUAL:
        return 8;

    // Bitwise operators
    case TokenType::AMPERSAND:
    case TokenType::PIPE:
    case TokenType::CARET:
        return 9;

    // Additive operators
    case TokenType::PLUS:
    case TokenType::MINUS:
        return 10;

    // Multiplicative operators
    case TokenType::STAR:
    case TokenType::SLASH:
    case TokenType::PERCENT:
        return 20;

    // Exponentiation operator
    case TokenType::POWERING:
        return 30;

    default:
        return 0;
    }
}

std::shared_ptr<ASTNode> Parser::parse_new_decl()
{
    bool isMutable = false;
    std::cout << "Before new Token\n";
    std::cout << peek().lexeme + "\n";

    const Token &newToken = consume(TokenType::NEW, "Expected 'new'");

    std::cout << "After new Token\n";
    std::cout << peek().lexeme + "\n";

    if (peek().type == TokenType::MUT)
    {
        isMutable = true;
        consume(TokenType::MUT, "Expected 'mut'");
    }

    std::cout << "After mut Token\n";
    std::cout << peek().lexeme + "\n";

    const Token &varName = consume(TokenType::IDENTIFIER, "Expected identifier for variable name");

    std::cout << "After varname\n";
    std::cout << peek().lexeme + "\n";

    consume(TokenType::COLON, "Expected ':' after variable name");

    std::cout << "After ':'\n";
    std::cout << peek().lexeme + "\n";

    const Token &typeToken = consume(TokenType::IDENTIFIER, "Expected type after ':'");

    std::cout << "After typeToken\n";
    std::cout << peek().lexeme + "\n";

    consume(TokenType::EQUAL, "Expected '=' after type");

    std::cout << "After '='\n";
    std::cout << peek().lexeme + "\n";

    auto value = expression(0); // Passar 0 como precedência inicial
    if (!value)
    {
        throw std::runtime_error("Expected expression after '='");
    }

    return std::make_shared<VarDeclarationNode>(varName, std::move(value), isMutable, newToken.loc);
}

File: src/parser/parser.ts
import { TypesNative } from "../compiler/values.ts";
import { Loc, Token, TokenType } from "../lexer/token.ts";
import {
  AST_FLOAT,
  AST_IDENTIFIER,
  AST_INT,
  AST_NULL,
  AST_STRING,
  BinaryExpr,
  Expr,
  Program,
  Stmt,
} from "./ast.ts";

type PrefixParseFn = () => Expr;
type InfixParseFn = (left: Expr) => Expr;

enum Precedence {
  LOWEST = 1,
  ASSIGN = 2, // =
  OR = 3, // ||
  AND = 4, // &&
  EQUALS = 5, // == !=
  COMPARISON = 6, // < > <= >=
  SUM = 7, // + -
  PRODUCT = 8, // * / %
  EXPONENT = 9, // **
  PREFIX = 10, // -x !x
  CALL = 11, // myFunction(x)
}

export class Parser {
  private tokens: Token[];
  private pos = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  // Entry point
  public parse(): Program {
    const program: Program = {
      kind: "Program",
      type: "null",
      value: null,
      body: [],
      loc: {} as Loc,
    };
    while (!this.isAtEnd()) {
      program.body!.push(this.parseStmt());
    }
    program.loc = this.tokens.length > 0
      ? this.tokens[0].loc && this.tokens[this.tokens.length - 1].loc
      : {} as Loc;
    return program;
  }

  private parseStmt(): Stmt {
    return this.parseExpressionStatement();
  }

  private parseExpressionStatement(): Expr {
    const expr = this.parseExpression(Precedence.LOWEST);
    // optionally consume semicolon
    if (this.match(TokenType.SEMICOLON)) {
      // ignore
    }
    return expr;
  }

  private parseExpression(precedence: Precedence): Expr {
    let left = this.parsePrefix();

    while (!this.isAtEnd() && precedence < this.peekPrecedence()) {
      const infix = this.getInfixFn(this.peek().kind);
      if (!infix) break;
      left = infix.call(this, left);
    }

    return left;
  }

  private parsePrefix(): Expr {
    const token = this.advance();

    switch (token.kind) {
      case TokenType.INT:
        return AST_INT(token.value as number, token.loc);
      case TokenType.FLOAT:
        return AST_FLOAT(token.value as number, token.loc);
      case TokenType.STRING:
        return AST_STRING(token.value as string, token.loc);
      case TokenType.NULL:
      case TokenType.EOF:
        return AST_NULL(token.loc);
      case TokenType.IDENTIFIER: {
        const name = token.value!.toString();
        if (this.peek().kind === TokenType.LPAREN) {
          return this.parseCallExpression(AST_IDENTIFIER(name, token.loc));
        }
        return AST_IDENTIFIER(name, token.loc);
      }
      case TokenType.LPAREN: {
        const expr = this.parseExpression(Precedence.LOWEST);
        this.consume(TokenType.RPAREN, "Expect ')' after expression.");
        return expr;
      }
      default:
        throw new Error(`No prefix parse function for ${token.value}`);
    }
  }

  private parseCallExpression(callee: Expr): Expr {
    // consume '('
    const paren = this.peek();
    this.advance();
    const args: Expr[] = [];
    if (!this.check(TokenType.RPAREN)) {
      do {
        args.push(this.parseExpression(Precedence.LOWEST));
      } while (this.match(TokenType.COMMA));
    }
    this.consume(TokenType.RPAREN, "Expect ')' after arguments.");
    return {
      kind: "CallExpr",
      callee,
      arguments: args,
      loc: this.makeLoc(callee.loc, paren.loc),
    } as any;
  }

  private getInfixFn(kind: TokenType): InfixParseFn | undefined {
    switch (kind) {
      case TokenType.PLUS:
      case TokenType.MINUS:
      case TokenType.SLASH:
      case TokenType.ASTERISK:
      case TokenType.EXPONENTIATION:
      case TokenType.PERCENT:
      case TokenType.REMAINDER:
      case TokenType.EQUALS_EQUALS:
      case TokenType.NOT_EQUALS:
      case TokenType.GREATER_THAN:
      case TokenType.LESS_THAN:
      case TokenType.GREATER_THAN_OR_EQUALS:
      case TokenType.LESS_THAN_OR_EQUALS:
      case TokenType.AND:
      case TokenType.OR:
        return this.parseBinaryInfix;
      default:
        return undefined;
    }
  }

  private parseBinaryInfix(left: Expr): Expr {
    this.advance();
    const operatorToken = this.previous();
    const precedence = this.getPrecedence(operatorToken.kind);
    const right = this.parseExpression(precedence);
    const type = this.inferType(left, right);
    return {
      kind: "BinaryExpr",
      operator: operatorToken.value!.toString(),
      left,
      right,
      type,
      loc: this.makeLoc(left.loc, right.loc),
    } as BinaryExpr;
  }

  // Helpers
  private isAtEnd(): boolean {
    return this.peek().kind === TokenType.EOF;
  }
  private peek(): Token {
    return this.tokens[this.pos];
  }
  private previous(): Token {
    return this.tokens[this.pos - 1];
  }
  private advance(): Token {
    if (!this.isAtEnd()) this.pos++;
    return this.previous();
  }
  private match(...kinds: TokenType[]): boolean {
    for (const kind of kinds) {
      if (this.check(kind)) {
        this.advance();
        return true;
      }
    }
    return false;
  }
  private check(kind: TokenType): boolean {
    if (this.isAtEnd()) return false;
    return this.peek().kind === kind;
  }
  private consume(expected: TokenType, message: string): Token {
    if (this.check(expected)) return this.advance();
    throw new Error(message);
  }
  private getPrecedence(kind: TokenType): Precedence {
    switch (kind) {
      case TokenType.EQUALS:
        return Precedence.ASSIGN;
      case TokenType.OR:
        return Precedence.OR;
      case TokenType.AND:
        return Precedence.AND;
      case TokenType.EQUALS_EQUALS:
      case TokenType.NOT_EQUALS:
        return Precedence.EQUALS;
      case TokenType.LESS_THAN:
      case TokenType.GREATER_THAN:
      case TokenType.LESS_THAN_OR_EQUALS:
      case TokenType.GREATER_THAN_OR_EQUALS:
        return Precedence.COMPARISON;
      case TokenType.PLUS:
      case TokenType.MINUS:
        return Precedence.SUM;
      case TokenType.SLASH:
      case TokenType.ASTERISK:
      case TokenType.PERCENT:
      case TokenType.REMAINDER:
        return Precedence.PRODUCT;
      case TokenType.EXPONENTIATION:
        return Precedence.EXPONENT;
      case TokenType.LPAREN:
        return Precedence.CALL;
      default:
        return Precedence.LOWEST;
    }
  }
  private peekPrecedence(): Precedence {
    return this.getPrecedence(this.peek().kind);
  }

  private inferType(left: Expr, right: Expr): TypesNative {
    if (left.type === "string" || right.type === "string") return "string";
    if (left.type === "float" || right.type === "float") return "float";
    return "int";
  }

  private makeLoc(start: Loc, end: Loc): Loc {
    return { ...start, end: end.end, line_string: start.line_string };
  }
}
